src/dtree.ml

```diff
@@ -0,0 +1,499 @@
+    rewritten to the action. *)
+module ClauseMat =
+struct
+  (** A component of the matrix, a term along with its position in the top
+      term. *)
+  type component = term * Subterm.t
+
+  (** [pp_component o c] prints component [c] to channel [o]. *)
+  let pp_component : component pp = fun oc (te, pos) ->
+    Format.fprintf oc "@[<h>%a@ %@@ %a@]" Print.pp te Subterm.pp pos
+
+  (** A redefinition of the rule type.
+
+      {b Note} that {!type:array} is used while {!module:ReductionStack} could
+      be used because {!val:pick_best_among} goes through all items of a rule
+      anyway ([max(S) = Θ(|S|)]).  Since heuristics need to access elements of
+      the matrix, we favour quick access with {!type:array}. *)
+  type rule =
+    { lhs : component array
+    (** Left hand side of a rule.   *)
+    ; rhs : action
+    (** Right hand side of a rule. *)
+    ; env_builder : (int * int) list
+    (** Maps an index int the {!val:vars} array to a slot. *) }
+
+  (** Type of a matrix of patterns.  Each line is a row having an attached
+      action. *)
+  type t =
+    { clauses : rule list
+    (** The rules. *)
+    ; var_met : int
+    (** Number of variables encountered so far. *) }
+
+  (** [pp o m] prints matrix [m] to out channel [o]. *)
+  let pp : t pp = fun oc { clauses ; _ } ->
+    let module F = Format in
+    let pp_line oc l =
+      F.fprintf oc "@[<h>" ;
+      F.pp_print_list ~pp_sep:(fun _ () -> F.fprintf oc ";@ ")
+        pp_component oc (Array.to_list l) ;
+      F.fprintf oc "@]" in
+    F.fprintf oc "{@[<v>@," ;
+    F.pp_print_list ~pp_sep:F.pp_print_cut pp_line oc
+      (List.map (fun { lhs ; _ } -> lhs) clauses) ;
+    F.fprintf oc "@.}@,"
+
+  (** [of_rules r] creates the initial pattern matrix from a list of rewriting
+      rules. *)
+  let of_rules : Terms.rule list -> t = fun rs ->
+    let r2r r =
+      let term_pos = List.to_seq r.Terms.lhs |> Subterm.tag |> Array.of_seq in
+      { lhs = term_pos ; rhs = r.Terms.rhs ; env_builder = [] } in
+    { clauses = List.map r2r rs ; var_met = 0 }
+
+  (** [is_empty m] returns whether matrix [m] is empty. *)
+  let is_empty : t -> bool = fun m -> m.clauses = []
+
+  (** [get_col n m] retrieves column [n] of matrix [m]. *)
+  let get_col : int -> t -> component list = fun ind { clauses ; _ } ->
+    List.map (fun { lhs ; _ } -> lhs.(ind)) clauses
+
+  (** [score c] returns the score heuristic for column [c]. *)
+  let rec score : component list -> int = function
+    | []                              -> 0
+    | (x, _) :: xs when is_treecons x -> score xs
+    | _ :: xs                         -> succ (score xs)
+
+  (** [pick_best_among m c] returns the index of the best column of matrix [m]
+      among columns [c] according to a heuristic. *)
+  let pick_best_among : t -> int array -> int = fun mat columns->
+    let wild_pc = Array.map (fun ci ->
+      let c = get_col ci mat in score c) columns in
+    Array.argmax (<=) wild_pc
```

Comment:
Can't you avoid the construction of lists (get_col) and the array (wild_pc) to compute argmax? You just need to go through the matrix and update the max score and column number.

Reply:
Eventually I'd like to revamp the way the heuristic is used, to have like a more "standard" approach